src/phoenix/config.py

```diff
@@ -621,11 +622,133 @@ def get_env_oauth2_settings() -> list[OAuth2ClientConfig]:
+
+        return connection_str
+    return None
+
+
+def _no_local_storage() -> bool:
+    """
+    Check if we're using a postgres database by checking if postgres connection string is set
+    and a working directory was not explicitly set.
+    """
+    return get_env_postgres_connection_str() is not None and getenv(ENV_PHOENIX_WORKING_DIR) is None
+
+
+class RestrictedPath(wrapt.ObjectProxy):  # type: ignore[misc]
+    def __init__(self, wrapped: Union[str, Path]) -> None:
+        super().__init__(Path(wrapped))
+
+    def _check_forbidden(self) -> None:
+        if _no_local_storage():
+            raise DirectoryError()
+        return
+
+    def __getattr__(self, name: str) -> Any:
+        attr = getattr(self.__wrapped__, name)
+
+        if callable(attr):
+
+            def wrapped_attr(*args: Any, **kwargs: Any) -> Any:
+                result = attr(*args, **kwargs)
+                if isinstance(result, Path):
+                    self._check_forbidden()
+                    return RestrictedPath(result)
+                elif hasattr(result, "__iter__") and not isinstance(result, (str, bytes)):
+                    return (RestrictedPath(p) if isinstance(p, Path) else p for p in result)
+                return result
+
+            return wrapped_attr
+        else:
+            if isinstance(attr, Path):
+                self._check_forbidden()
+                return RestrictedPath(attr)
+            return attr
+
+    def __str__(self) -> str:
+        self._check_forbidden()
+        return str(self.__wrapped__)
+
+    def __repr__(self) -> str:
+        return f"<RestrictedPath({repr(self.__wrapped__)})>"
+
+    def __fspath__(self) -> str:
+        self._check_forbidden()
+        return str(self.__wrapped__)
+
+    def __truediv__(self, other: Union[str, Path]) -> "RestrictedPath":
+        new_path = self.__wrapped__ / other
+        return RestrictedPath(new_path)
+
+    def __itruediv__(self, other: Union[str, Path]) -> "RestrictedPath":
+        self.__wrapped__ /= other
+        self._check_forbidden()
+        return self
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, RestrictedPath):
+            return bool(self.__wrapped__ == other.__wrapped__)
+        return bool(self.__wrapped__ == other)
+
+    def __hash__(self) -> int:
+        return hash(self.__wrapped__)
+
+    def __iter__(self) -> Iterator["RestrictedPath"]:
+        self._check_forbidden()
+        for child in self.__wrapped__:
+            yield RestrictedPath(child)
+
+    def __len__(self) -> int:
+        return len(self.__wrapped__.parts)
+
+    def __contains__(self, item: str) -> bool:
+        return item in self.__wrapped__.parts
+
+
+ROOT_DIR = RestrictedPath(WORKING_DIR)
+EXPORT_DIR = RestrictedPath(ROOT_DIR / "exports")
+INFERENCES_DIR = RestrictedPath(ROOT_DIR / "inferences")
+TRACE_DATASETS_DIR = RestrictedPath(ROOT_DIR / "trace_datasets")
+
+
 def ensure_working_dir() -> None:
```

Comment:
```suggestion
def ensure_working_dir_if_needed() -> None:
```

Suggestion, non-blocking